vnext/Microsoft.ReactNative.Cxx/JSValue.h

```diff
@@ -19,119 +19,438 @@ IJSValueReader MakeJSValueTreeReader(JSValue &&root) noexcept;
+  uint64_t AsUInt64() const noexcept;
+
+  //! Return a Double representation of JSValue. It is the same as JavaScript Number(value) result.
+  double AsDouble() const noexcept;
+
+  //! Return an Float representation of JSValue. It is the same as JavaScript Number(value) result casted to float.
+  float AsFloat() const noexcept;
+
+  //! Return value T that is created from JSValue using the ReadValue function override.
+  //! Default T is constructed by using default constructor.
+  template <
+      class T,
+      std::enable_if_t<std::is_default_constructible_v<T> && !std::is_constructible_v<T, std::nullptr_t>, int> = 0>
+  T As() const noexcept;
+
+  //! Return value T that is created from JSValue using the ReadValue function override.
+  //! Default T is constructed by using constructor that receives nullptr as a parameter.
+  template <class T, std::enable_if_t<std::is_constructible_v<T, std::nullptr_t>, int> = 0>
+  T As() const noexcept;
+
+  //! Return value T that is created from JSValue using the ReadValue function override.
+  //! Default T is constructed from the provided 'default' value.
   template <class T>
-  static JSValue From(const T &value) noexcept;
+  T As(T &&defaultValue) const noexcept;
 
-  const JSValue &GetObjectProperty(std::string_view propertyName) const noexcept;
-  const JSValue &GetArrayItem(JSValueArray::size_type index) const noexcept;
-  const JSValue &operator[](std::string_view propertyName) const noexcept;
-  const JSValue &operator[](JSValueArray::size_type index) const noexcept;
+  //! Convert JSValue to a readable string that can be used for logging.
+  std::string ToString() const noexcept;
 
-  bool Equals(const JSValue &other) const noexcept;
+#pragma endregion
 
+#pragma region JSValue serialization and deserialization.
+
+  //! Create JSValue from IJSValueReader.
   static JSValue ReadFrom(IJSValueReader const &reader) noexcept;
-  static JSValueObject ReadObjectFrom(IJSValueReader const &reader) noexcept;
-  static JSValueArray ReadArrayFrom(IJSValueReader const &reader) noexcept;
 
+  //! Write this JSValue to IJSValueWriter.
   void WriteTo(IJSValueWriter const &writer) const noexcept;
-  static void WriteObjectTo(IJSValueWriter const &writer, JSValueObjectView object) noexcept;
-  static void WriteArrayTo(IJSValueWriter const &writer, JSValueArrayView value) noexcept;
 
- private:
-  bool ObjectEquals(const JSValueObject &other) const noexcept;
-  bool ArrayEquals(const JSValueArray &other) const noexcept;
-  static JSValueObject ReadObjectProperties(IJSValueReader const &reader) noexcept;
-  static JSValueArray ReadArrayItems(IJSValueReader const &reader) noexcept;
+#pragma endregion
+
+#pragma region Deprecated methods
+
+  // The methods below are deprecated in favor of other methods with clearer semantic
+  [[deprecated("Use TryGetObject or AsObject")]] JSValueObject const &Object() const noexcept;
```

Comment:
I'm guessing you're aware, but depending on warning level this can present as a compiler error. Forcing someone to suppress the deprecation is probably good though.

Reply:
My initial intent was to delete these functions in favor of new one, but I realized that in some cases developers will have no idea what to do. The deprecated attribute gives me an excellent tool to communicate which function developers must use instead of the old ones.

---
In reply to: [389199820](https://github.com/microsoft/react-native-windows/pull/4260#discussion_r389199820) [](ancestors = 389199820)